icu4c/source/common/uassert.h

```diff
@@ -10,25 +10,53 @@
 *
 * File uassert.h
 *
-*  Contains U_ASSERT macro
-*
-*    By default, U_ASSERT just wraps the C library assert macro.
-*    By changing the definition here, the assert behavior for ICU can be changed
-*    without affecting other non-ICU uses of the C library assert().
+*  Contains the U_ASSERT and UPRV_UNREACHABLE macros
 *
 ******************************************************************************
 */
-
 #ifndef U_ASSERT_H
 #define U_ASSERT_H
+
 /* utypes.h is included to get the proper define for uint8_t */
 #include "unicode/utypes.h"
+/* platform.h is included for U_CPLUSPLUS_VERSION */
+#include "unicode/platform.h"
+
+/**
+ * \def U_ASSERT
+ * By default, U_ASSERT just wraps the C library assert macro.
+ * By changing the definition here, the assert behavior for ICU can be changed
+ * without affecting other non - ICU uses of the C library assert().
+*/
 #if U_DEBUG
 #   include <assert.h>
 #   define U_ASSERT(exp) assert(exp)
 #else
 #   define U_ASSERT(exp)
 #endif
-#endif
 
+/**
+ * \def UPRV_UNREACHABLE
+ * This macro is used to unconditionally abort if unreachable code is ever executed.
+ * @internal
+*/
+#if defined(UPRV_UNREACHABLE)
+    // Use the predefined value.
+#else
+#   if defined(__cplusplus) && U_CPLUSPLUS_VERSION >= 11
+    // For C++11 code, we can use the C++11 attribute [[noreturn]] to indicate to the
+    // compiler that execution never returns from this function.
+    [[noreturn]] inline void
+    uprv_abort_noreturn();
+        // For C code, and C++ below C++11, we need to use compiler specific annotations.
+#   elif defined(_MSC_VER)
+        inline void __declspec(noreturn) uprv_abort_noreturn();
+#   else
+        inline void __attribute__((noreturn)) uprv_abort_noreturn();
+#   endif
+#   include <stdlib.h>
```

Comment:
Should the definition of uprv_abort_noreturn() be separated from that of UPRV_UNREACHABLE? So that it doesn't go away if someone provides their own UPRV_UNREACHABLE. Someone may want to use it separately.

The stdlib abort() function is already declared noreturn by the language specification, so we could probably shorten ours to uprv_abort(); adding noreturn to the name feels redundant.

The # include <stdlib.h> should probably go up top, with other #includes.

Reply:
Thanks Andy.

When I was looking at [CppReference ](https://en.cppreference.com/w/c/program/abort) it looked like the stdlib `abort()` function was only defined noreturn in the newer C11 standard, and I don't think ICU requires C11 currently. (That's why I did it the way it is with the `uprv_abort_noreturn()` wrapper function.)
However, when I look at the MSVC headers, even though MSVC doesn't fully support C11 it seems that abort is already defined as noreturn, so we might actually be able to simplify this to just calling `abort()`.

Do we want to have `uprv_abort_noreturn()` (or `uprv_abort()`) be its own function? (I don't have strong feelings either way).

Good point about the include for stdlib. I'll move the include up with the other includes.

Thanks again.

Reply:
I'll push another commit that uses just `abort()`, so we can see how the CI builds are and if there are any warnings.  (Though if you want I can also bring back the uprv_abort_noreturn() or uprv_abort() function too).

Reply:
Good point about plain C and abort(). We don't require C11, so uprv_abort_noreturn() makes sense. Feel free to put it back if you prefer it - I'm happy with it either way.

Reply:
I think I will leave it just calling `abort()` directly for now, since that seems simpler/cleaner, and I also don't see any warnings in the CI build outputs about noreturn.
I suppose if some other compilers besides GCC, Clang, and MSVC have issues we can always revise it and bring back the helper/wrapper function uprv_abort().
(I'll squash the 3 commits into 1 comment and then force push).

Thanks for taking a look and for the feedback Andy!